File: serial_exercise/java/src/test/java/SerialTest.java
package com.softlayer.labs.interview;
import java.util.List;

import junit.framework.TestCase;

public class SerialTest extends TestCase
{
    public void testSimpleSequentialList() {
        Serial serial = new Serial();
        List serials = serial.generate("a2010", 10, 1);

        assertEquals(10, serials.size());
        assertEquals("a2010", serials.get(0));
        assertEquals("a2011", serials.get(1));
        assertEquals("a2012", serials.get(2));
        assertEquals("a2013", serials.get(3));
        assertEquals("a2014", serials.get(4));
        assertEquals("a2015", serials.get(5));
        assertEquals("a2016", serials.get(6));
        assertEquals("a2017", serials.get(7));
        assertEquals("a2018", serials.get(8));
        assertEquals("a2019", serials.get(9));
    }
    
    public void testSimpleOffsetList() {
        Serial serial = new Serial();
        List serials = serial.generate("a2010", 10, 2);

        assertEquals(10, serials.size());
        assertEquals("a2010", serials.get(0));
        assertEquals("a2012", serials.get(1));
        assertEquals("a2014", serials.get(2));
        assertEquals("a2016", serials.get(3));
        assertEquals("a2018", serials.get(4));
        assertEquals("a201a", serials.get(5));
        assertEquals("a201c", serials.get(6));
        assertEquals("a201e", serials.get(7));
        assertEquals("a201g", serials.get(8));
        assertEquals("a201i", serials.get(9));
    }
    
    public void testSmallNumber() {
        Serial serial = new Serial();
        List serials = serial.generate("0", 11, 4);

        assertEquals(11, serials.size());
        assertEquals("0", serials.get(0));
        assertEquals("4", serials.get(1));
        assertEquals("8", serials.get(2));
        assertEquals("c", serials.get(3));
        assertEquals("g", serials.get(4));
        assertEquals("k", serials.get(5));
        assertEquals("o", serials.get(6));
        assertEquals("s", serials.get(7));
        assertEquals("w", serials.get(8));
        assertEquals("10", serials.get(9));
        assertEquals("14", serials.get(10));
    }
}

File: serial_exercise/go/serial_test.go
package main

import (
	"errors"
	"testing"
)

func verifyResults(t *testing.T, result []string, expectedResult []string) {
	if len(result) != len(expectedResult) {
		t.Errorf("Wrong number of serials returned. got: %d, want: %d", len(result), len(expectedResult))
	}

	if len(expectedResult) > 0 {
		for i, serial := range result {
			if expectedResult[i] != serial {
				t.Errorf("Unexpected serial at %d. got: %s, want: %s.", i, serial, expectedResult[i])
			}
		}
	}
}

func verifyError(t *testing.T, err error, expectedErr error) {
	if err == nil && expectedErr != nil {
		t.Errorf("Expected error not returned. want: %s", expectedErr.Error())
		return
	}

	if err != nil && expectedErr == nil {
		t.Errorf("Unexpected error returned. got: %s", err.Error())
		return
	}

	if err != nil && expectedErr != nil && err.Error() != expectedErr.Error() {
		t.Errorf("Returned error does not match expected error. got: %s, want: %s", err.Error(), expectedErr.Error())
	}
}

func TestEmptyList(t *testing.T) {
	result, err := generateSerialNumbers("", 10, 1)

	expectedResult := []string{}

	verifyError(t, err, errors.New("Invalid startSerial"))
	verifyResults(t, result, expectedResult)
}

func TestSimpleSequentialList(t *testing.T) {
	result, err := generateSerialNumbers("a2010", 10, 1)

	expectedResult := []string{
		"a2010",
		"a2011",
		"a2012",
		"a2013",
		"a2014",
		"a2015",
		"a2016",
		"a2017",
		"a2018",
		"a2019",
	}

	verifyError(t, err, nil)
	verifyResults(t, result, expectedResult)
}

func TestSimpleOffsetList(t *testing.T) {
	result, err := generateSerialNumbers("a2010", 10, 2)

	expectedResult := []string{
		"a2010",
		"a2012",
		"a2014",
		"a2016",
		"a2018",
		"a201a",
		"a201c",
		"a201e",
		"a201g",
		"a201i",
	}

	verifyError(t, err, nil)
	verifyResults(t, result, expectedResult)
}

func TestBadCount(t *testing.T) {
	result, err := generateSerialNumbers("a2010", -1, 1)

	expectedResult := []string{}

	verifyError(t, err, errors.New("Invalid count"))
	verifyResults(t, result, expectedResult)
}

func TestBadOffset(t *testing.T) {
	result, err := generateSerialNumbers("a2010", 10, -1)

	expectedResult := []string{}

	verifyError(t, err, errors.New("Invalid offset"))
	verifyResults(t, result, expectedResult)
}
